### PR TITLE
Git and provider website redirects

### PR DESCRIPTION
**Pull request by elceebee**
### Context

We have improved our course pages and we need a way of monitoring how our users are interacting with them

[Trello card](https://trello.com/c/y3NKamG9)

### Changes proposed in this pull request
The redirect allows us to capture the initial click in big query before sending people out to the provider or get-into-teaching websites.

### Guidance to review
We can only tests that the redirects take us where we ultimately want to go. We will only know that they are being registered in BigQuery in production.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally
- [x] Inform data insights team due to database changes

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -856,6 +856,7 @@ en:
               taken: "Provider code already taken"
             provider_type:
               blank: "Provider type can't be blank"
+              format: Accredited provider cannot be a school
```
Comment (elceebee) [inline config/locales/en.yml:859]: Unrelated change - just removed the duplicate key, `provider_type`

File: app/controllers/concerns/provider_website_redirect.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+module ProviderWebsiteRedirect
+  extend ActiveSupport::Concern
+
+  def provider_website
+    redirect_to provider.decorate.website, allow_other_host: true
```
Comment (CatalinVoineag) [inline app/controllers/concerns/provider_website_redirect.rb:7]: I assume you need to have this controller that will actually redirect to where the user wants because we only capture requests to our own controllers with dfe_analytics? We don't capture requests to outside services?

**Review by CatalinVoineag — APPROVED** [review]
Nice work 👍🏻

File: app/controllers/concerns/provider_website_redirect.rb (hunk 2 above)
Comment (elceebee) [reply app/controllers/concerns/provider_website_redirect.rb:7]: Yeah, that's exactly right. We do this currently with the apply button on the course pages as well, I think it is a nice little solution without having to set up beacons or fire off other events (it looks like @tomas-stefano set it up for the apply button). The analysts will be able to sort for clicks to `course/:provider_code/:course_code:/provider_url` for instance, and know how many people visited a provider's website.

Comment (CatalinVoineag) [reply app/controllers/concerns/provider_website_redirect.rb:7]: Interesting, thank you 👍🏻

Comment (tomas-stefano) [reply app/controllers/concerns/provider_website_redirect.rb:7]: Awesome idea. 🚀

**Review by tomas-stefano — APPROVED** [review]

File: config/routes/find.rb [hunk 3]
```diff
@@ -15,6 +15,8 @@
   get '/course/:provider_code/:course_code', to: 'courses#show', as: 'course'
   get '/course/:provider_code/:course_code/placements', to: 'placements#index', as: :placements
   get '/course/:provider_code/:course_code/apply', to: 'courses#apply', as: :apply
+  get '/course/:provider_code/:course_code/git/:git_path', to: 'courses#git_redirect', as: :get_into_teaching_redirect
+  get '/course/:provider_code/:course_code/provider_website', to: 'courses#provider_website', as: :provider_website
```
Comment (dcyoung-dev) [inline config/routes/find.rb]:
```suggestion
  get '/course/:provider_code/:course_code/provider/website', to: 'courses#provider_website', as: :provider_website
```

Not a massive suggestion but thought this would leave the door open for adding other links related to the provider.

File: app/controllers/concerns/get_into_teaching_redirect.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  extend ActiveSupport::Concern
+
+  def git_redirect
+    redirect_to I18n.t("get_into_teaching.#{git_url_locale_key}"), allow_other_host: true
+  end
+
+  private
+
+  def git_path_param
+    params[:git_path]
```
Comment (dcyoung-dev) [inline app/controllers/concerns/get_into_teaching_redirect.rb:13]: should we validate the params we accept?

If we put something silly as the param where do we get redirected to? i.e. where would `course/:provider_code/:course_code/git/help` redirect to?

**Review by dcyoung-dev — COMMENTED** [review]
A quick sanity check on the params accepted for the git endpoints

File: app/controllers/concerns/get_into_teaching_redirect.rb (hunk 4 above)
Comment (elceebee) [reply app/controllers/concerns/get_into_teaching_redirect.rb:13]: That will fail, but I think that's ok. It should fail if we don't have a locale. I thought about catching the failure and directing to the git home page, but that would hide the error.

**Review by dcyoung-dev — APPROVED** [review]
